### PR TITLE
Add RIOT examples

### PR DESCRIPTION
**Pull request by LasseRosenow**
The only relevant changes here are in the "examples" directory.

Probably not the most clean solution, but it works for now.

See my comments for some questions.

### CODE REVIEW

File: examples/riot/external_modules/reactor-uc/Makefile.include
Comment (LasseRosenow) [inline examples/riot/external_modules/reactor-uc/Makefile.include:8]: Regarding this thing I have no Idea why I needed to do it like this, but it works for now. I will ask some people who know the riot build system better to improve this in the future ..

File: examples/riot/src/blinky/Makefile
Comment (LasseRosenow) [inline examples/riot/src/blinky/Makefile:8]: I have specified the RIOT folder to be one higher thant the root of this project. So it is easy to just clone and put the RIOT folder next to this project.

Maybe we can improve the ergonomix here by including RIOT as a submodule?
If this is the way to go, I was not yet sure where to put it. Should we put the RIOT submodule just in the examples/riot folder?

**Review by erlingrj — CHANGES_REQUESTED** [review]
Happy that you are tackling this. A question that we should find an answer to right away is how we should integrate with RIOT. Tassilo started on something in his fork of RIOT (https://github.com/tanneberger/RIOT), but we got hold up because I couldn't make the RIOT make-based build system do what I wanted. 

In Tassilos fork, we added reactor-uc as an external package by pointing to a commit of this repo. We could then write a RIOT programming using the reactor-uc package and linking it. Currently we are thinking of LF as the language for writing the applications so we would have LFC code-generating a project that uses reactor-uc. Normally, LFC also copies the entire reactor-uc library into the project. So I am not sure whether it makes sense to have reactor-uc as an external package in RIOT or just part of each project.

What do you think @tanneberger?

File: examples/riot/src/blinky/Makefile
Comment (LasseRosenow) [reply examples/riot/src/blinky/Makefile:8]: Yes including this project into riot as a package is easy and a common way to do it.
But for testing this project on riot and doing active development on it I find the RIOT package system inkonvenient. It requires you to point to a specific commit ID and we always need to update the RIOT package to new versions.

Thats why I added this example within this project. So I see 2 options: 
- Option 1: Add RIOT as a submodule to the examples/riot folder, which does not mean this project depends on RIOT, but just makes it more easy for new people to run the riot examples in here.
- Option 2: Keep as is and just explain in the readme that one needs to clone RIOT next to this project on their file system to run the examples.

Both ways are totally reasonable for me, but feel free to come up with an even better solution :);)

File: examples/riot/src/blinky/Makefile
Comment (LasseRosenow) [reply examples/riot/src/blinky/Makefile:24]: I agree. I was not sure on how hard reactor-uc's dependency on ztimer and ztimer_usec is. If you think it is always needed then I would move it up to the external_modules makefile

File: examples/riot/external_modules/reactor-uc/Makefile
Comment (LasseRosenow) [reply examples/riot/external_modules/reactor-uc/Makefile:4]: I personally prefer not to have so much platform specific files in the root folder, but if you think this is fine then sure we can move the make files for riot into the root folder. This can then also be reused in package of reactor-uc that is included into RIOT OS in @tanneberger's fork :)

File: examples/riot/src/blinky/Makefile
Comment (erlingrj) [reply examples/riot/src/blinky/Makefile:8]: I prefer 2 but I would like to not rely on relative paths, what about requiring an environmental variable, e.g. `RIOT_PATH` or `REACTOR_UC_RIOT_PATH` or something?

File: examples/riot/src/blinky/Makefile
Comment (erlingrj) [reply examples/riot/src/blinky/Makefile:24]: It is a key dependency, it is what we use to implement that `wait_until` function which puts the runtime to sleep until an absolute time-point in the future

File: examples/riot/src/blinky/Makefile
Comment (LasseRosenow) [reply examples/riot/src/blinky/Makefile:8]: In the marked line you can see that there already is an environment varialble: "RIOTBASE" this is the common variable name we use everywhere in RIOT OS applications. The relative path is just a fallback if that variable is not specified

File: examples/riot/external_modules/reactor-uc/Makefile
Comment (LasseRosenow) [reply examples/riot/external_modules/reactor-uc/Makefile:4]: Okay then I would propose to put RIOT as a submodule into `external/platform/RIOT` for now.

> we need to add nanopb sources to the build. This file can be used by the RIOT build system, or by another Makefile-based build system if the user wants to integrate it with something different.

Yes this seems reasonable I will see to move the non-riot-specific parts to the root folder.

> I was really struggling with getting the RIOT make system to do what I wanted

The same for me :) The RIOT build system can be quite tough and confusing. I myself are still quite confused about how exactly that Makefile.include file exactly works... I hope I can make the reactor-uc external module better soon.

File: examples/riot/src/blinky/Makefile
Comment (LasseRosenow) [reply examples/riot/src/blinky/Makefile:24]: I had some issues including it in the external_module. For some reason the makefile.include file ignores it if i place the USEMODULEs there. But I am confident I can find a fix soon

Comment (LasseRosenow) [reply examples/riot/src/blinky/Makefile]: This is fixed now. I these dependencies to the `Makefile.dep` file of the `reactor-uc` `external_module`

**Comment by LasseRosenow** [comment]
Okay I reverted the RIOT submodule :)

**Comment by LasseRosenow** [comment]
I set this to draft because even though the threads work now, there are still some compilation problems with riot because of stricter rules. I will see to disable them somehow in make.

File: examples/riot/blinky/main.c
Comment (tanneberger) [inline examples/riot/blinky/main.c:4]: if you rebase master you can use the new macro shorthands.

Comment (LasseRosenow) [reply examples/riot/blinky/main.c]: Done

**Comment by LasseRosenow** [comment]
@erlingrj I think this should be ready to merge now. Feel free to comment if you find remaining problems with this PR :)

**Review by erlingrj — CHANGES_REQUESTED** [review]
Great! I am able to run the blinky example with the `native` "board". However, I would prefer not to add those TcpIpChannel tests as examples because it duplicates everything and making changes to TcpIpChannel will require changes all over.

In fact, I would like to see TcpIpChannel examples moved to the unit-test folder instead.

So I am happy to merge this if we can remove the TcpIpChannel stuff from RIOT examples

**Comment by LasseRosenow** [comment]
Okay I have rebased with main and fixed the issues.
I have removed the tcp test examples and added a basic hello example just like zephyr has.

File: src/platform/posix/tcp_ip_channel.c
Comment (LasseRosenow) [inline src/platform/posix/tcp_ip_channel.c:221]: I needed to do these changed to the tcp_ip_channel because the RIOT POSIX wrappers don't support `pthread_attr_setstack`. I will try to fix that on RIOT ASAP but for now I added this `define` to support both cases.
It is needed to do it like this because `pthread_attr_setstackaddr` is deprecated. So we only want to use it if not avoidable.

Comment (erlingrj) [reply src/platform/posix/tcp_ip_channel.c:221]: OK, we should figure it out, the reason I added it was because on Zephyr you have to do it that way...

**Comment by erlingrj** [comment]
@LasseRosenow thanks for this, great work. I am merging it and I will resolve the failing tests in #71

**Pull request merged**
